contracts/poap/src/contract.rs

```diff
@@ -0,0 +1,1086 @@
+    use crate::test_utils::{
+        get_valid_init_msg, EVENT_END_SECONDS, EVENT_START_SECONDS, INITIAL_BLOCK_TIME_SECONDS,
+    };
+    use crate::ContractError;
+    use crate::ContractError::Unauthorized;
+    use cosmwasm_std::testing::{mock_dependencies, mock_env, mock_info};
+    use cosmwasm_std::{DepsMut, Timestamp};
+
+    const CREATOR: &str = "creator";
+    const ADMIN: &str = "admin";
+    const MINTER: &str = "minter";
+    const USER: &str = "user";
+
+    fn do_instantiate(deps: DepsMut) {
+        let mut env = mock_env();
+        let info = mock_info(CREATOR, &vec![]);
+
+        // Change block time to event start.
+        env.block.time = Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS);
+
+        let msg = get_valid_init_msg(1);
+        assert!(instantiate(deps, env, info, msg).is_ok());
+    }
+
+    #[test]
+    fn instantiate_with_invalid_admin_addr() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+
+        let mut init_msg = get_valid_init_msg(1);
+        init_msg.admin = "a".to_string();
+
+        let result = instantiate(deps.as_mut(), env, info, init_msg);
+        assert!(result.is_err())
+    }
+
+    #[test]
+    fn instantiate_with_invalid_minter_addr() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+
+        let mut init_msg = get_valid_init_msg(1);
+        init_msg.minter = "a".to_string();
+
+        let result = instantiate(deps.as_mut(), env, info, init_msg);
+        assert!(result.is_err())
+    }
+
+    #[test]
+    fn instantiate_with_invalid_creator_addr() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+
+        let mut init_msg = get_valid_init_msg(1);
+        init_msg.event_info.creator = "a".to_string();
+
+        let result = instantiate(deps.as_mut(), env, info, init_msg);
+        assert!(result.is_err())
+    }
+
+    #[test]
+    fn instantiate_with_event_start_before_current_time() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+
+        let mut init_msg = get_valid_init_msg(1);
+        // Create a start time 200 seconds before the current block time
+        let start = &env.block.time.seconds() - 200;
+        init_msg.event_info.start_time = Timestamp::from_seconds(start);
+        init_msg.event_info.end_time = Timestamp::from_seconds(start + 600);
+
+        let init_result = instantiate(deps.as_mut(), env.clone(), info, init_msg);
+        assert_eq!(
+            ContractError::StartTimeBeforeCurrentTime {
+                current_time: env.block.time,
+                start_time: Timestamp::from_seconds(start)
+            },
+            init_result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_event_start_equal_current_time() {
```

Comment:
It is duplicated with `instantiate_with_event_start_before_current_time`. It can be replaced into one test which current block time is equal to start time which is considered as `before current time`.

Reply:
This ensure that also that the time equal branch is covered instead of just the before. I do agree that is very similar to the other but is just another test to esure that when we receive a time that is equal to the current one the contract handles it like is a time before the current time.